File: src/Services/WorkCalendar.php
<?php

declare(strict_types=1);

namespace App\Services;

use App\Support\DateTimeHelper;
use DateInterval;
use DateTimeImmutable;
use RuntimeException;

final class WorkCalendar
{
    public function __construct(
        private readonly array $intervals,
        private readonly array $workingDays = [1, 2, 3, 4, 5],
        private readonly array $holidays = [],
        private readonly array $dayOrdersByDate = []
    ) {
        if ($intervals === []) {
            throw new RuntimeException('Nenhum intervalo de trabalho cadastrado.');
        }
    }

    private function isOrderAllowedForDay(int $orderNumber, DateTimeImmutable $day): bool
    {
        $dateKey = $day->format('Y-m-d');

        // Regra A: se a data existir em orders_by_day, ela é soberana (sem fallback).
        if (array_key_exists($dateKey, $this->dayOrdersByDate)) {
            $dayOrders = $this->dayOrdersByDate[$dateKey] ?? null;
            if (!is_array($dayOrders)) {
                return false;
            }

            $hasAnyAllowedOrder = false;
            for ($order = 1; $order <= 4; $order++) {
                if (!empty($dayOrders[$order])) {
                    $hasAnyAllowedOrder = true;
                    break;
                }
            }

            if (!$hasAnyAllowedOrder) {
                return false;
            }

            return !empty($dayOrders[$orderNumber]);
        }

        // Data ausente: fallback antigo (ordem não é filtrada por data).
        return true;
    }

    public function nextValidDateTime(DateTimeImmutable $dateTime): DateTimeImmutable
    {
        $currentInterval = $this->findCurrentInterval($dateTime);

        if ($currentInterval !== null) {
            return $dateTime;
        }

        for ($offset = 0; $offset <= 30; $offset++) {
            $day = $dateTime->setTime(0, 0)->add(new DateInterval('P' . $offset . 'D'));

            if (!$this->hasIntervalsForDay($day)) {
                continue;
            }

            foreach ($this->intervalInstancesForDay($day) as $interval) {
                if ($dateTime < $interval['start']) {
                    return $interval['start'];
                }
            }
        }

        throw new RuntimeException('Nao foi possivel encontrar o proximo horario valido.');
    }

    public function addWorkingMinutes(DateTimeImmutable $start, int $minutes): DateTimeImmutable
    {
        return $this->buildWorkingPlan($start, $minutes)['end'];
    }

    public function buildWorkingPlan(DateTimeImmutable $start, int $minutes): array
    {
        $current = $this->nextValidDateTime($start);
        $remaining = $minutes;
        $segments = [];

        while ($remaining > 0) {
            $interval = $this->findCurrentInterval($current);

            if ($interval === null) {
                $current = $this->nextValidDateTime($current);
                continue;
            }

            $available = (int) floor(($interval['end']->getTimestamp() - $current->getTimestamp()) / 60);

            if ($available <= 0) {
                $current = $this->nextValidDateTime($interval['end']);
                continue;
            }

            $consumed = min($remaining, $available);
            $segmentEnd = DateTimeHelper::addMinutes($current, $consumed);

            $segments[] = [
                'start' => $current,
                'end' => $segmentEnd,
                'minutes' => $consumed,
                'interval_start' => $interval['start'],
                'interval_end' => $interval['end'],
                'interval_minutes' => (int) floor(($interval['end']->getTimestamp() - $interval['start']->getTimestamp()) / 60),
            ];

            if ($remaining <= $available) {
                return [
                    'end' => $segmentEnd,
                    'segments' => $segments,
                ];
            }

            $remaining -= $consumed;
            $current = $this->nextValidDateTime($interval['end']);
        }

        return [
            'end' => $current,
            'segments' => $segments,
        ];
    }

    public function workingMinutesBetween(DateTimeImmutable $start, DateTimeImmutable $end): int
    {
        if ($end <= $start) {
            return 0;
        }

        $minutes = 0;
        $cursor = $start;

        while ($cursor < $end) {
            $validCursor = $this->nextValidDateTime($cursor);

            if ($validCursor >= $end) {
                break;
            }

            $interval = $this->findCurrentInterval($validCursor);

            if ($interval === null) {
                break;
            }

            $segmentEnd = $interval['end'] < $end ? $interval['end'] : $end;
            $minutes += (int) floor(($segmentEnd->getTimestamp() - $validCursor->getTimestamp()) / 60);
            $cursor = $segmentEnd;
        }

        return $minutes;
    }

    private function hasIntervalsForDay(DateTimeImmutable $day): bool
    {
        return $this->intervalInstancesForDay($day) !== [];
    }

    private function holidayDateKeys(): array
    {
        $keys = [];

        foreach ($this->holidays as $holiday) {
            if (is_array($holiday)) {
                $date = (string) ($holiday['date'] ?? '');
            } else {
                $date = (string) $holiday;
            }

            $date = trim($date);
            if ($date !== '') {
                $keys[] = $date;
            }
        }

        return array_values(array_unique($keys));
    }

    private function isCalendarOpenForDay(DateTimeImmutable $day): bool
    {
        return !in_array($day->format('Y-m-d'), $this->holidayDateKeys(), true);
    }

    private function isIntervalAllowedForDay(array $interval, DateTimeImmutable $day): bool
    {
        if (!$this->isCalendarOpenForDay($day)) {
            return false;
        }

        $dayNumber = (int) $day->format('N');
        $intervalDays = array_values(array_filter(
            array_map('intval', $interval['days'] ?? $this->workingDays),
            static fn (int $value): bool => $value >= 1 && $value <= 7
        ));

        return in_array($dayNumber, $intervalDays, true);
    }

    private function findCurrentInterval(DateTimeImmutable $dateTime): ?array
    {
        $previousDay = $dateTime->setTime(0, 0)->sub(new DateInterval('P1D'));
        foreach ($this->intervalInstancesForDay($previousDay) as $interval) {
            if ($dateTime >= $interval['start'] && $dateTime < $interval['end']) {
                return $interval;
            }
        }

        $currentDay = $dateTime->setTime(0, 0);
        foreach ($this->intervalInstancesForDay($currentDay) as $interval) {
            if ($dateTime >= $interval['start'] && $dateTime < $interval['end']) {
                return $interval;
            }
        }

        return null;
    }

    private function intervalInstancesForDay(DateTimeImmutable $day): array
    {
        if (!$this->isCalendarOpenForDay($day)) {
            return [];
        }

        $dateKey = $day->format('Y-m-d');
        $hasDayOrders = array_key_exists($dateKey, $this->dayOrdersByDate);
        $allowedOrdersForDate = $hasDayOrders && is_array($this->dayOrdersByDate[$dateKey] ?? null)
            ? $this->dayOrdersByDate[$dateKey]
            : [];

        if ($hasDayOrders) {
            $hasAnyAllowedOrder = false;
            for ($order = 1; $order <= 4; $order++) {
                if (!empty($allowedOrdersForDate[$order])) {
                    $hasAnyAllowedOrder = true;
                    break;
                }
            }

            // Data presente na agenda + todas as ordens desmarcadas = dia bloqueado (sem fallback).
            if (!$hasAnyAllowedOrder) {
                return [];
            }
        }

        $instances = [];

        foreach ($this->intervals as $index => $interval) {
            $orderNumber = isset($interval['order']) ? (int) $interval['order'] : ($index + 1);

            if ($hasDayOrders) {
                if (empty($allowedOrdersForDate[$orderNumber])) {
                    continue;
                }
            }

            if (!$this->isIntervalAllowedForDay($interval, $day)) {
                continue;
            }

            [$startHour, $startMinute] = array_map('intval', explode(':', $interval['start']));
            [$endHour, $endMinute] = array_map('intval', explode(':', $interval['end']));

            $start = $day->setTime($startHour, $startMinute);
            $end = $day->setTime($endHour, $endMinute);

            if ($end <= $start) {
                $nextDay = $day->add(new DateInterval('P1D'));
                $nextDayStart = $nextDay->setTime(0, 0);
                $end = $nextDay->setTime($endHour, $endMinute);

                // Regra B: trecho após 00:00 só existe se o dia seguinte for válido para a mesma ordem.
                // Regra B (ajustada): turno overnight pertence ao dia em que iniciou (bloqueia domingo/feriado).
                $nextDayNumber = (int) $nextDay->format('N');
                if ($nextDayNumber === 7 || !$this->isCalendarOpenForDay($nextDay)) {
                    $end = $nextDayStart;
                }
            }

            $instances[] = [
                'start' => $start,
                'end' => $end,
            ];
        }

        usort(
            $instances,
            static fn (array $left, array $right): int => $left['start'] <=> $right['start']
        );

        return $instances;
    }
}

File: python_gantt/app/services/work_calendar.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timedelta
from typing import Any


@dataclass(frozen=True)
class CalendarInterval:
    start: str
    end: str
    days: list[int]
    order: int


class WorkCalendar:
    def __init__(
        self,
        intervals: list[dict[str, Any]],
        working_days: list[int] | None = None,
        holidays: list[dict[str, Any] | str] | None = None,
    ) -> None:
        if not intervals:
            raise ValueError("Nenhum intervalo de trabalho cadastrado.")

        self.intervals = intervals
        self.working_days = working_days or [1, 2, 3, 4, 5]
        self.holidays = holidays or []

    @staticmethod
    def _parse_datetime(value: datetime | str) -> datetime:
        if isinstance(value, datetime):
            return value

        raw = str(value or "").strip()
        if raw == "":
            raise ValueError("Data/hora invalida.")

        raw = raw.replace("T", " ")
        formats = [
            "%Y-%m-%d %H:%M:%S",
            "%Y-%m-%d %H:%M",
            "%Y-%m-%d",
            "%d/%m/%Y %H:%M:%S",
            "%d/%m/%Y %H:%M",
            "%d/%m/%Y",
        ]
        for fmt in formats:
            try:
                return datetime.strptime(raw, fmt)
            except ValueError:
                continue

        return datetime.fromisoformat(raw)

    @staticmethod
    def _minutes_between(start: datetime, end: datetime) -> int:
        if end <= start:
            return 0
        return int((end - start).total_seconds() // 60)

    def _holiday_date_keys(self) -> list[str]:
        keys: list[str] = []
        for holiday in self.holidays:
            if isinstance(holiday, dict):
                date_value = str(holiday.get("date") or "").strip()
            else:
                date_value = str(holiday).strip()
            if date_value:
                keys.append(date_value)
        return list(dict.fromkeys(keys))

    def _is_calendar_open_for_day(self, day: datetime) -> bool:
        return day.strftime("%Y-%m-%d") not in self._holiday_date_keys()

    def _is_interval_allowed_for_day(self, interval: dict[str, Any], day: datetime) -> bool:
        if not self._is_calendar_open_for_day(day):
            return False

        interval_days = interval.get("days") or self.working_days
        allowed_days = [int(value) for value in interval_days if 1 <= int(value) <= 7]
        return day.isoweekday() in allowed_days

    def _interval_instances_for_day(self, day: datetime) -> list[dict[str, datetime]]:
        if not self._is_calendar_open_for_day(day):
            return []

        instances: list[dict[str, datetime]] = []
        for index, interval in enumerate(self.intervals):
            if not self._is_interval_allowed_for_day(interval, day):
                continue

            start_raw = str(interval.get("start") or "").strip()
            end_raw = str(interval.get("end") or "").strip()
            if start_raw == "" or end_raw == "":
                continue

            start_hour, start_minute = [int(part) for part in start_raw.split(":")[:2]]
            end_hour, end_minute = [int(part) for part in end_raw.split(":")[:2]]

            start = day.replace(hour=start_hour, minute=start_minute, second=0, microsecond=0)
            end = day.replace(hour=end_hour, minute=end_minute, second=0, microsecond=0)

            if end <= start:
                next_day = day + timedelta(days=1)
                end = next_day.replace(hour=end_hour, minute=end_minute, second=0, microsecond=0)
                if next_day.isoweekday() == 7 or not self._is_calendar_open_for_day(next_day):
                    end = next_day.replace(hour=0, minute=0, second=0, microsecond=0)

            instances.append({"start": start, "end": end, "order": index + 1})

        instances.sort(key=lambda item: item["start"])
        return instances

    def _find_current_interval(self, date_time: datetime) -> dict[str, datetime] | None:
        previous_day = (date_time.replace(hour=0, minute=0, second=0, microsecond=0) - timedelta(days=1))
        for interval in self._interval_instances_for_day(previous_day):
            if interval["start"] <= date_time < interval["end"]:
                return interval

        current_day = date_time.replace(hour=0, minute=0, second=0, microsecond=0)
        for interval in self._interval_instances_for_day(current_day):
            if interval["start"] <= date_time < interval["end"]:
                return interval

        return None

    def next_valid_datetime(self, date_time: datetime) -> datetime:
        current_interval = self._find_current_interval(date_time)
        if current_interval is not None:
            return date_time

        for offset in range(31):
            day = date_time.replace(hour=0, minute=0, second=0, microsecond=0) + timedelta(days=offset)
            if not self._interval_instances_for_day(day):
                continue

            for interval in self._interval_instances_for_day(day):
                if date_time < interval["start"]:
                    return interval["start"]

        raise ValueError("Nao foi possivel encontrar o proximo horario valido.")

    def working_minutes_between(self, start: datetime, end: datetime) -> int:
        if end <= start:
            return 0

        minutes = 0
        cursor = start

        while cursor < end:
            valid_cursor = self.next_valid_datetime(cursor)
            if valid_cursor >= end:
                break

            interval = self._find_current_interval(valid_cursor)
            if interval is None:
                break

            segment_end = interval["end"] if interval["end"] < end else end
            minutes += self._minutes_between(valid_cursor, segment_end)
            cursor = segment_end

        return minutes

    def consolidate_intervals(self, intervals: list[dict[str, datetime]]) -> list[dict[str, datetime]]:
        if not intervals:
            return []

        ordered = sorted(intervals, key=lambda item: (item["start"], item["end"]))
        merged = [ordered[0].copy()]

        for interval in ordered[1:]:
            current = merged[-1]
            if interval["start"] <= current["end"]:
                if interval["end"] > current["end"]:
                    current["end"] = interval["end"]
                continue
            merged.append(interval.copy())

        return merged
